example/Disparity.hs

```diff
@@ -0,0 +1,93 @@
+{-# LANGUAGE AllowAmbiguousTypes, CPP, DataKinds, FlexibleInstances, GADTs,
+             QuantifiedConstraints, RankNTypes, TypeFamilies #-}
+{-# OPTIONS_GHC -Wno-incomplete-patterns  #-}
+{-# OPTIONS_GHC -Wno-missing-export-lists #-}
+{-# OPTIONS_GHC -Wno-compat-unqualified-imports #-}
+{-# OPTIONS_GHC -Wno-missing-signatures #-}
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
+
+module Disparity where
+import Prelude
+import GHC.TypeLits
+import HordeAd.Core.DualNumber
+import HordeAd.Core.DualClass           as DC
+import HordeAd
+import qualified System.Random          as R
+import qualified Data.Array.ShapedS     as OS
+import qualified Foreign.Storable       as F
+import Data.List
+
+
+-- | Disparity cost volume.
+--
+--   Take two arrays of multi channel 2d images, where the first contains
+--   left views of the scene and the second contains right views.
+--
+--   For each pair of images, slice the right image over the left image,
+--   and for each offset produce the L1 distance indicating how well correponding
+--   multi-channel image elements in the right image match those in the left.
+--
+--   Described in:
+--    Anytime Stereo Image Depth Estimation on Mobile Devices
+--    Wang, Lai et al, ICRA 2019
+--    https://arxiv.org/abs/1810.11408
+--    Section III b).
+--
+costVolume
+ :: forall d r nImgs nChas nRows nCols nCount shA shO.
+    ( ADModeAndNum d r
+    , KnownNat nImgs, KnownNat nChas, KnownNat nRows, KnownNat nCols, KnownNat nCount
```

Comment:
For the one that has `StaticNat`, you don't need the `KnownNat` any more. The rule of thumb seems to be, if the compiler is giving you grief or if the function is expected to be used by other users, not just by your other function, then do `StaticNat` for all dimension arguments.